### PR TITLE
DNM : Race condition in rados bench

### PR DESCRIPTION
**Pull request by Abhishekvrshny**
http://tracker.ceph.com/issues/12949

### CODE REVIEW

**Comment by dachary** [comment]
@Abhishekvrshny the link in the description must be an issue in the "Backport" tracker

**Comment by Abhishekvrshny** [comment]
@dachary Yes my bad :( . It should have been 12949. Can I change that somehow?

**Comment by dachary** [comment]
@Abhishekvrshny I see you figured it out ;-)

**Comment by Abhishekvrshny** [comment]
@dachary https://github.com/ceph/ceph/pull/5152 was merged into master after https://github.com/ceph/ceph/pull/4690 was merged. Can that be the reason for failures? Shall we wait for https://github.com/ceph/ceph/pull/5810 to get into hammer first before backporting this particular PR ?

**Comment by dachary** [comment]
You can include fa72ecb and  	dba8b5b from #5810 and see if it helps. I would do:

    git reset --hard ceph/hammer
    git cherry-pick fa72ecb dba8b5b be2aab5
    git push --force

and see if the bot compiles ok with that. It's perfectly ok to include commits from other pull requests if needed. They will be silently trimmed by git when merging.

**Comment by shinobu-x** [comment]
@dachary very nice introduction not for bug-fix ... Sorry it's not quite about this pr.

File: src/common/obj_bencher.cc [hunk 1]
```diff
@@ -153,8 +153,12 @@ void *ObjBencher::status_printer(void *_bencher) {
 
 int ObjBencher::aio_bench(
   int operation, int secondsToRun,
+<<<<<<< HEAD
```
Comment (theanalyst) [inline src/common/obj_bencher.cc]: These lines shouldn't be present in the commits..

File: src/common/obj_bencher.cc [hunk 2]
```diff
@@ -153,8 +153,12 @@ void *ObjBencher::status_printer(void *_bencher) {
...
   int maxObjectsToCreate,
   int concurrentios, int op_size, bool cleanup, const char* run_name) {
+=======
+  int concurrentios, int object_size, bool cleanup, const char* run_name, bool no_verify) {
+>>>>>>> ca6abca... tools: add --no-verify option to rados bench
```
Comment (theanalyst) [inline src/common/obj_bencher.cc]: I'm assuming the second version of the line ie: the line with bool no verfiy should be the only line present.. line 158 should be deleted

File: src/common/obj_bencher.h [hunk 3]
```diff
@@ -64,9 +64,15 @@ class ObjBencher {
 
   int fetch_bench_metadata(const std::string& metadata_file, int* object_size, int* num_objects, int* prevPid);
 
+<<<<<<< HEAD
   int write_bench(int secondsToRun, int maxObjects, int concurrentios, const string& run_name_meta);
   int seq_read_bench(int secondsToRun, int num_objects, int concurrentios, int writePid);
   int rand_read_bench(int secondsToRun, int num_objects, int concurrentios, int writePid);
+=======
```
Comment (theanalyst) [inline src/common/obj_bencher.h]: again .. line 67-71 shouldnt be present.. also line 75

File: src/common/obj_bencher.cc [hunk 4]
```diff
@@ -797,16 +846,24 @@ int ObjBencher::rand_read_bench(int seconds_to_run, int num_objects, int concurr
     lock.Lock();
     ++data.started;
     ++data.in_flight;
-    snprintf(data.object_contents, data.object_size, "I'm the %16dth object!", current_index);
+<<<<<<< HEAD
```
Comment (theanalyst) [inline src/common/obj_bencher.cc]: delete this too..

**Comment by Abhishekvrshny** [comment]
@theanalyst my bad. The side-effect of manually merging conflicts

**Comment by theanalyst** [comment]
@Abhishekvrshny no problems... this is a non trivial merge .. and a good learning experience

**Comment by dachary** [comment]
@Abhishekvrshny I suggest you try to get all modifications mande to rados bench since hammer with
<pre>
git cherry-pick -x 069d95eaf49cadaa9a8fa1fa186455944a50ec7d e360bfd7058af01fd03aa9260eacbdc5d631e34b ddb422f40394ec2f4fc4a6a6cff49e43538121a9 39442641fc61c6ba7bde2cde5ceafd9276647f85 e3d62a9d4c6da7bd48f7aefce91db050bdc9a637 31d16e90108a741e92f897e550a8bf1e26392d2d 1d1d0aa01b944e8f9dad96b7f13bce2a581ba758 a87ac4d36b86aa826afb990769831bfcf08ed3fa cd0f2b754dc3cdf1e4f1647dc1cd91d891ce896c 1ac727982213bf676beefe9340d6822193dbb700 5c6190d012785222ba1eebee734aa34f79c5bf27 fe79babc9867a4cdf81a5b83019b9e1bbc86f9ae b894fc7acf7dee7f7ec8c5c280e7a6be41133328 ca6abca 55a6f9e b894fc7 9bcf5f0
</pre>
It applies cleanly and it's legitimate IMHO because:

* it fixes a few bugs
* it is limited to the obj_bencher.{cc,h} and rados.cc files
* a few cleanups that occurred in between bug fixes make it difficult to backport cleanly
* although it's made of a dozen commit the series is straightforward

 <pre>
a417478 tools: fix race condition in seq/rand bench
 src/common/obj_bencher.cc | 80 ++++++++++++++++++++++++++++++++++++++++++++------------------------------------
 1 file changed, 44 insertions(+), 36 deletions(-)
697cfc9 bufferlist: implement bufferlist::invalidate_crc()
 src/common/buffer.cc   |  9 +++++++++
 src/include/buffer.h   |  1 +
 src/test/bufferlist.cc | 40 ++++++++++++++++++++++++++++++++++++++++
 3 files changed, 50 insertions(+)
8c3d381 tools: add --no-verify option to rados bench
 src/common/obj_bencher.cc | 67 ++++++++++++++++++++++++++++++++++++++++++-------------------------
 src/common/obj_bencher.h  |  6 +++---
 src/tools/rados/rados.cc  | 11 ++++++++++-
 3 files changed, 55 insertions(+), 29 deletions(-)
ba2bf47 tools: Don't delete, recreate and re-fill buffers in rados bench.
 src/common/obj_bencher.cc | 32 +++++++++++++++++++-------------
 1 file changed, 19 insertions(+), 13 deletions(-)
056ecdb common: fix ObjBencher::aio_bench signature
 src/common/obj_bencher.cc | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)
036744d common: rados bench data.finished = 0
 src/common/obj_bencher.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)
193fa52 common: remove unused maxObjectsToCreate
 src/common/obj_bencher.cc | 22 +++++++++-------------
 src/common/obj_bencher.h  |  4 ++--
 src/tools/rados/rados.cc  |  2 +-
 3 files changed, 12 insertions(+), 16 deletions(-)
b128b78 obj_bencher: does not accumulate bandwidth that is zero
 src/common/obj_bencher.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
f3d718a obj_bencher: aio_bench - rename op_size to object_size
 src/common/obj_bencher.cc | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)
bd2d94d obj_bencher: remove excess 'object_size = op_size'
 src/common/obj_bencher.cc | 2 --
 1 file changed, 2 deletions(-)
e1797ca obj_bencher: remove 'trans_size' as obsolete
 src/common/obj_bencher.cc | 5 ++---
 src/common/obj_bencher.h  | 1 -
 2 files changed, 2 insertions(+), 4 deletions(-)
13b3290 common: make rados bench return correctly errno.
 src/common/obj_bencher.cc | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)
2566198 obj_bencher: add IOPS metric calculation
 src/common/obj_bencher.cc | 28 ++++++++++++++++++++++++++++
 src/common/obj_bencher.h  |  3 +++
 2 files changed, 31 insertions(+)
2188183 obj_bencher: cosmetic display fixes
 src/common/obj_bencher.cc | 30 ++++++++++++------------------
 1 file changed, 12 insertions(+), 18 deletions(-)
455b7f8 obj_bencher: generalize vec_stddev function
 src/common/obj_bencher.cc | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)
8813436 obj_bencher: fix indents
 src/common/obj_bencher.cc | 116 ++++++++++++++++++++++++++++++++++++++++++++++++------------------------------------------------
 1 file changed, 58 insertions(+), 58 deletions(-)
</pre>

**Comment by Abhishekvrshny** [comment]
@dachary This would put a bulk of commits in one backport PR. Is that fine? or shall there be separate backport PRs for each of them?

**Comment by dachary** [comment]
> Also if we take all changes since hammer, how do we know if there isn't something which is NOT meant to be in hammer (and is probably a feature only for the next release)?

Very good question :-) The only way is to review each commit individually, which I did. I guess we're lucky that this is a simple situation. It may have been a lot more hairy if there were new features.

**Comment by dachary** [comment]
> This would put a bulk of commits in one backport PR. Is that fine? 

This is generally not ok and we do that very rarely. IMHO it's ok in this specific case because it's about a tool and not the core and because the series of change is quite simple.

> or shall there be separate backport PRs for each of them? 

Let's first see how it goes with tests. If tests are happy and noone objects this rather extensive backport, then we can worry about cross referencing this commit in the right places.

You're lucky : you got to deal with one of the most complicated cases of backport :-)

**Comment by Abhishekvrshny** [comment]
@dachary :) . Now that the tests have passed, shall I make a new branch with an appropriate name and send a separate PR, or will this be fine?

**Comment by branch-predictor** [comment]
I didn't expected that it'll be so complicated. Sorry, guys!
And thanks for the hard work.

**Comment by dachary** [comment]
@branch-predictor it is a good experience (not just for @Abhishekvrshny ;-).

**Comment by dachary** [comment]
@Abhishekvrshny I think it's worth asking if https://github.com/ceph/ceph/pull/5428 should be backported (on the ceph-devel mailing list).

**Comment by dachary** [comment]
For the record, the mail thread is at : http://permalink.gmane.org/gmane.comp.file-systems.ceph.devel/27302

**Comment by dachary** [comment]
<pre>
<BranchPredictor> abhishekvrshny: why we need #5428? #5812 relies on it?
<BranchPredictor> ah, right
<BranchPredictor> there's a change to rest_bench as well, but it got lost once rest_bench was eradicated
<BranchPredictor> abhishekvrshny: this: http://ceph.predictor.org.pl/diff-wip-12947.txt should help
<abhishekvrshny> BranchPredictor: Can you raise a hammer specific PR for 5812 and I would close 5812 itself. I can then schedule an integration test with that.
<BranchPredictor> abhishekvrshny: well, I can...
<BranchPredictor> company will kill me for all those git clones. :P
<abhishekvrshny> BranchPredictor: it looks like a non-trivial backport to me. I dont want to mess it up :P 
<BranchPredictor> abhishekvrshny: no worries, I'll do it.
<abhishekvrshny> BranchPredictor: Thanks :)
</pre>

**Comment by branch-predictor** [comment]
125456 < BranchPredictor> abhishekvrshny: why we need #5428? #5812 relies on it?
125545 < abhishekvrshny> loicd: http://permalink.gmane.org/gmane.comp.file-systems.ceph.devel/27302
125606 < BranchPredictor> ah, right
125631 < BranchPredictor> there's a change to rest_bench as well, but it got lost once rest_bench was eradicated
125729 < BranchPredictor> abhishekvrshny: this: http://ceph.predictor.org.pl/diff-wip-12947.txt should help
130233 < abhishekvrshny> BranchPredictor: Can you raise a hammer specific PR for 5812 and I would close 5812 itself. I can then schedule an integration test with that.
130305 < BranchPredictor> abhishekvrshny: well, I can...
130423 < abhishekvrshny> BranchPredictor: it looks like a non-trivial backport to me. I dont want to mess it up :P
130444 < BranchPredictor> abhishekvrshny: no worries, I'll do it.
131010 < abhishekvrshny> BranchPredictor: Thanks :)

**Comment by dachary** [comment]
replaced  by https://github.com/ceph/ceph/issues/6791

**Pull request closed without merging**